Add check for cert signature algo matches the negotiated algo

Wenxing-hou opened the pull request:
Fix: #492 

**Patch1: Add check for cert signature algo matches the negotiated algo:**
- Get the cert signature algo oid.
- Get the libspdm defined signature_algo oid by negotiated algo.
- Compare the cert_signature_algo oid with libspdm defined signature_algo.

**Patch2: Add basic constraints check for leaf cert:**
- Check will pass for two case:

> 1. basic constraints is not present in cert;
> 2. leaf cert basic_constraints CA is false;

**Patch3: Add Hardware Identity OID check for AliasCert model cert:**
- Move x509_get_key_usage() before x509_get_extended_key_usage;
- Check if Hardware Identity OID present in AliasCert model leaf cert:

> 1. Find the spdm extensions OID;
> 2. Find Hardware Identity OID;
> 3. Check if conflict for AliasCert model;
> 4. Add unit test and update cert;


Signed-off-by: Wenxing Hou <wenxing.hou@intel.com>

jyao1 commented on library/spdm_crypt_lib/libspdm_crypt_crypt.c:
```diff
@@ -3864,16 +3899,213 @@ static bool internal_spdm_x509_date_time_check(IN uint8_t *from,
+
+    /*sm2 oid  TBD*/
+    case SPDM_ALGORITHMS_BASE_ASYM_ALGO_TPM_ALG_SM2_ECC_SM2_P256:
+        ASSERT(false);
```
this is will cause current SM2 cert in EMU fail.

I think we need make it pass temporarily and add SM2 specific check later.

Wenxing-hou replied:
Yes. Because the libspdm_get_signature_algo_OID function only is called by libspdm_verify_cert_signature_algo_OID now.
And in libspdm_verify_cert_signature_algo_OID, I have added the code before:
```
    /*work around: skip the sm2*/
    if (base_asym_algo == SPDM_ALGORITHMS_BASE_ASYM_ALGO_TPM_ALG_SM2_ECC_SM2_P256) {
        return true;
    }
```
I think your mean is the libspdm_get_signature_algo_OID function may be called by other function in the future.
I will fix the issue.

jyao1 commented:
See feedback on https://github.com/DMTF/libspdm/issues/678

Wenxing-hou commented:
> See feedback on #678

Thanks. I have updated the code with the feedback. And I have added the SM2withSM3 OID.
So, this PR have been completed all.

jyao1 commented:
> Thanks. I have updated the code with the feedback. And I have added the SM2withSM3 OID.
> So, this PR have been completed all.

I suggest to drop SM2, if it is not tested yet.

Wenxing-hou commented on library/spdm_common_lib/libspdm_com_crypto_service.c:
```diff
@@ -592,10 +592,19 @@ bool libspdm_verify_peer_cert_chain_buffer(libspdm_context_t *spdm_context,
     uintn received_root_cert_size;
     bool result;
     uint8_t root_cert_index;
+    bool Is_DeviceCert_model;
```
Yes. I have changed the variable name. And the SM2 patch has been dropped in this PR.

jyao1 commented on include/hal/library/cryptlib.h:
```diff
@@ -2800,6 +2800,52 @@ return_status x509_get_extended_key_usage(const uint8_t *cert,
+ * @retval RETURN_NOT_FOUND         If no Extension entry match oid.
+ * @retval RETURN_UNSUPPORTED       The operation is not supported.
+ **/
+return_status x509_get_extended_spdm_oid(const uint8_t *cert,
```
This function shall be moved to spdm_crypto_lib instead of crypto_lib.

jyao1 commented on include/hal/library/cryptlib.h:
```diff
@@ -2800,6 +2800,29 @@ return_status libspdm_x509_get_extended_key_usage(const uint8_t *cert,
+ * @retval RETURN_NOT_FOUND         If no Extension entry match oid.
+ * @retval RETURN_UNSUPPORTED       The operation is not supported.
+ **/
+return_status x509_get_extended_basic_constraints(const uint8_t *cert,
```
please follow the naming conversion. we need libspdm_ prefix

steven-bellock commented on library/spdm_requester_lib/libspdm_req_get_certificate.c:
```diff
@@ -219,7 +223,7 @@ return_status libspdm_try_get_certificate(void *context, uint8_t slot_id,
             spdm_context, libspdm_get_managed_buffer(&certificate_chain_buffer),
             libspdm_get_managed_buffer_size(&certificate_chain_buffer),
-            trust_anchor, trust_anchor_size);
+            trust_anchor, trust_anchor_size, is_requester);
```
Can use a `true` literal and remove `is_requester` variable.

Wenxing-hou replied:
Thanks. I have changed the code to `true`.

steven-bellock commented on library/spdm_responder_lib/libspdm_rsp_encap_get_certificate.c:
```diff
@@ -194,7 +198,7 @@ return_status libspdm_process_encap_response_certificate(
             libspdm_get_managed_buffer_size(
                 &spdm_context->encap_context.certificate_chain_buffer),
-            NULL, NULL);
+            NULL, NULL, is_requester);
```
Can use a `false` literal and remove `is_requester` variable.

Wenxing-hou replied:
Thanks. I have changed the code to `false`.

steven-bellock commented on os_stub/cryptlib_mbedtls/pk/x509.c:
```diff
@@ -19,9 +19,15 @@
+#define OID_EXT_KEY_USAGE     { 0x55, 0x1D, 0x25 }
+#define OID_BASIC_CONSTRAINTS { 0x55, 0x1D, 0x13 }
+
+static uint8_t m_libspdm_oid_common_name[] = OID_COMMON_NAME;
```
Can these also be `const`?

Wenxing-hou replied:
Yes. I have added the `const`.

jyao1 commented:
feedback is addressed. Merge.
